### PR TITLE
Martial Arts Rebalance for Bionic Combatives

### PR DESCRIPTION
**Pull request by Hymore246**
#### Summary

```SUMMARY: Balance "Martial arts rebalance for Bionic Combatives"```

#### Purpose of change

Partially implements #32422
Update for the martial art: Bionic Combatives

#### Describe the solution

`"A modern combat style for the post-modern human.  Nicknamed "Biojutsu", Bionic Combatives combines integrated weaponry, armor and augments into an consolidated fighting discipline.`

Bionic Combatives is martial art that only exists in CDDA and doesn't have any backstory associated with it. I've written a lore backstory for the martial art at the end of this PR in order to try and fill that gap.

_Current Buffs, Techniques, and Special Abilities_

Name | Level Available | Type | Effect 
--- | --- | --- | --- 
Arm Block|Unarmed 0||Needs Arms Alloy Plating CBM
Leg Block|Unarmed 0||Needs Legs Alloy Plating CBM
Biojutsu|Melee 2|Static Buff|+1 Blocks
biojutsu counter|Melee 4|Block Counter|
quick punch|Unarmed 0|Unarmed Tech|Rapid
quick slash|Melee 0|Melee Tech|Rapid
biojutsu impale|Melee 3|Melee Crit Tech|+50% Bash and Cut damage, Stun duration: 1
sweep kick|Melee 3|Unarmed Tech|Down duration: 2
wide strike|Melee 5|Melee Crit Tech|AoE: Wide

**Weapons**: Bionic Claws, Monomolecular Blade

Bionic Combatives is an interesting style. You can't learn it with a starting trait, NPCs don't teach it (and shouldn't), and no book exists that can teach it. Only the Close Quarters Battle CBM can allow you to use it. It's possible to permanently learn it from the CBM by using the style but it takes awhile. In the end is it worth the trouble? Not really. 

_Problems and issues_
- **Doesn't offer much when unarmed.**
Obviously, Bionic Combatives should focus on using bionic weapons but it's unarmed side doesn't really offer much to players that haven't found the weapon CBMs. 

- **Biojutsu Impale doesn't affect stabbing damage.**
This is a problem because stab damage is the main source of damage for [Bionic Claws.](http://cdda-trunk.chezzo.com/bio_claws_weapon) 

- **Almost good at AoE.**
Bionic Combatives is ALMOST good at AoE fighting but is lacking in a few places.

_Rebalance Goal_
Since Bionic Combatives is unique to CDDA and doesn't have a defined history in-game, I decided to approach the style by focusing on the following points:
- The style can be used both unarmed and armed but has mostly different techniques for each.
- It has niche at being the only offensive AoE style.
- Think like a computer.

In the case of point 3, computers are quick and precise in what they do. I decided to take a meta approach to the style by only giving it techniques that have conditional requirements. For the non-programmers reading, a conditional statement is used by programming languages to perform different actions depending on if something is True or False. A basic example of this would be:

```
if( I_AM_A_STATEMENT == true ) {
    print( "The statement is true" );
} else {
    print( "The statement is false" );
}
```

If the case of Bionic Combatives, it's decision tree of new techniques looks like this:

Drawio biojutsu

The above is not accurate 100% of the time due to the random nature of how CDDA picks techniques. It's possible to Measured Strike instead of Disarm or Biojutsu Takedown and Biojutsu Impale instead of Biojutsu Cleave. In these cases, I've added weighting to Disarm, Biojutsu Takedown, and Biojutsu Cleave to increase the chances of them occurring.

_New Buffs, Techniques, and Special Abilities_

Name | Level Available | Type | Effect 
--- | --- | --- | --- 
Arm Block|||Needs Arms Alloy Plating CBM
Leg Block|||Needs Legs Alloy Plating CBM
Biojutsu Stance|Melee 0|Static Buff|+2 Blocks attempts
Measured Strike (melee)|Melee 0|Melee Tech|80% move cost
Measured Strike (unarmed)|Unarmed 0|Unarmed Crit OK Tech|80% move cost
Disarm|Unarmed 1|Unarmed Crit OK Tech|Disarm, Stun duration: 1, weighting: 2
Biojutsu Takedown|Unarmed 2|Unarmed Crit OK Tech|Down duration: 2, weighting: 2
Biojutsu Impale|Melee 3|Melee Crit Tech|+50% Bash, Cut, Stab damage, Stun duration: 1
Biojutsu Counter|Melee 4|Block Counter|
Optimization|Unarmed 4|OnKill Buff|+2 Bash, Cut and Stab damage, +1 Accuracy. Lasts 3 turns, stacks 3 times.
Biojutsu Cleave|Melee 5|Melee Crit Tech|AoE: Wide, weighting: 2

**Weapons**: Bionic Claws, Monomolecular Blade

_Important Changes_
- **Biojutsu Impale increases stab damage by +50%**
No reason not to do this. Bionic Claws needs this change and "impaling" usually refers to something sharp.

- **Added weighting to several techs**
As mentioned above, I've added weighting to Disarm, Biojutsu Takedown, and Biojutsu Cleave to increase the chances of them occuring.

- **Measured Strike replaces Quick Punch/Quick Slash**
Unless a style is completely based around speed, it's usually a bad idea to have a Rapid technique as the only non-crit technique in a style. While it's higher damage overall (50% move cost / 66% damage done = 132% average damage) it also is less per attack which makes it worse against high armor enemies. So I replaced it with Measured Strike which is 80% move cost which is technically less damage overall (120% vs 132%) but better damage per attack.

- **There are two version of Measured Strike.**
This was done to make sure the Measured Strike only triggers it absolutely has to. For unarmed, it's a normal or crit technique that triggers when nothing else will. For Melee it's only a normal technique to force Biojutsu Impale to trigger on a critical hit unless you are attacking a group.

- **Added Optimization OnKill buff.**
Much like a computer, once your done with a task, move on and complete another task as efficiently as possible. Optimization is buff that assists getting kills quicker. It has a short duration but can help in a pinch. It can stack but due to the fact that buff stacks decay at a rate of 1 per turn (regardless of duration), you will not be seeing Optimization stack unless you are literally killing one or more things per turn.

_Advantages_
- The conditional attributes make Bionic Combatives very reliable in what technique you are going to get when you attack.
- This is the only style in the game with an AoE technique. With enough skill, it is extremely useful in killing weaker zombies en mass.
- Having +2 Block attempts and block counter is very useful for handling groups as well.

_Disadvantages_
- You need a bionic weapon to get the most damage and crowd control from the martial art.
- You need the Alloy Plating arms or legs CBMs if you want to block and counter while unarmed.
- Bionic Combatives is the hardest martial art to learn. You can only get it through the Close Quarters Battle CBM which locks your skill level at 5 when used. It's possible to permanently learn a martial art through the CQB CBM but the changes are really low. 
FORMULA (line 1313, melee.cpp):
```
    //player has a very small chance, based on their intelligence, to learn a style whilst using the CQB bionic
    if( has_active_bionic( bio_cqb ) && !has_martialart( style_selected ) ) {
        /** @EFFECT_INT slightly increases chance to learn techniques when using CQB bionic */
        // Enhanced Memory Banks bionic doubles chance to learn martial art
        const int bionic_boost = has_active_bionic( bionic_id( bio_memory ) ) ? 2 : 1;
        if( one_in( ( 1400 - ( get_int() * 50 ) ) / bionic_boost ) ) {
            ma_styles.push_back( style_selected );
            add_msg_if_player( m_good, _( "You have learned %s from extensive practice with the CQB Bionic." ),
                               _( style_selected.obj().name ) );
        }
    }
```

#### Describe alternatives you've considered

- I did not balanced Bionic Combatives based on how rare it is. I was extremely tempted to do so but this is a bad idea.  Since the rarity of the style could change at some point in the future, Bionic Combatives could become unbalanced if it's power was based on being rare.
- My original design for Bionic Combatives was to make the unarmed side disable enemies and the melee side take advantage of the disabled state to do more damage. This would require the player to transition between armed and unarmed to get the most out of the style. This ended up being extremely tedious because you would have to change back and forth very frequently. It also cost too much bionic power to do this and I also couldn't justify a way for the martial art to reduce the power cost for bionic weapon to 0 to allow this work. It also meant that the martial art wouldn't work without having a bionic weapon. 

_If the style is too strong..._
- Reduce the damage of Biojutsu Impale to +25%.
- Change the move cost of Measured Strike to 90%.
- Reduce Optimization damage from 2 to 1.
- Remove Accuracy buff from Optimization.

_If the style is too weak..._
- Adding another passive that increases unarmed bash damage.
- Increase the damage done by Biojutsu Counter to +25%.

#### Additional context

_Backstory_
With the emergence of Compact Bionic Modules (CBMs), life as the world had known it changed forever. The ability to modify the human body with machines sparked worldwide interest and the United States military was among the most interested. Not long after US military research into CBMs began, China reveal their own version of CBMs and their first wave of “steel soldiers.” This immediately sparked a bionic arms race between the two countries that continued for years. 

One of the more note worthy bionics created during this time came from a man named Dr. Paul Baker. Dr. Baker and his team managed to create a CBM that could control a person’s movements and actions by binding the bionic to the subject’s nervous system. This bionic came to be known as the Close Quarters Battle CBM and its primary function was to allow soldiers to fight like highly skilled martial artist without any real training. The CBM  was extremely well received not only for its effectiveness in combat but also the possibility of applying the technology to other skills and activities. However, tensions were beginning to run high in the arms race between the USA and China and pressure was put on Dr. Baker to improve the CBM with the goal of making every soldier a grandmaster of every martial art. Dr. Baker and his team soon discovered this to be an impossible task. 

Loading the CBM with too much information and forcing perfect movement had catastrophic results. Too much data overwhelmed the bionic and affected subjects would begin to move extremely slowly. In other tests, soldiers would break there own limbs by moving in impossible ways. In the worst cases, the CBM could cause damage to the user’s brain and nerves which usually ended in paralysis and death. Unable to halt the development due to the mounting pressure from the military, Dr. Baker came up with a solution to the problem. Instead of mastering multiple martial arts, why not master one perfect martial art? Calling in a few favors, Dr. Baker gained access to a supercomputer network designed to conduct military combat simulations. Supplying it with data on as many martial arts as he could find, he instructed the computer develop the perfect martial art for bionic soldiers.

After days of computations and simulations, the supercomputer network finished with an unexpected result. The supercomputer indicated that the data given to it was insufficient to create the desired martial art. However, it had produced a prototype martial art for the purpose of collecting combat data. With this data, a superior combat style could be created. Military leadership caught wind of this result and demanded that this "Bionic Combatives" prototype be put to immediately use. The prototype was given to soldiers that were deploying to handle the sudden rise in violent riots around the United States. The results from these bionic operators were both spectacular and terrifying. The martial art was described as being "completely relentless", “horribly efficient”, and “heartless”. Pleased with the results, military leadership ordered that their top soldiers to be given the prototype. These elite operators, outfitted with the most advanced CBMs available, were sent out into the field for testing. Dr. Baker and his team waited anxiously for the results, but none ever came. The Cataclysm struck and life as the world had known it changed, once again, forever.

TL:DR
- The Close Quarters Battle CBM was created Dr. Baker for the United States military.
- Bionic Combatives was created by a supercomputer network. It is actually a field prototype designed to gather combat data for the creation of the ultimate martial art for bionic super soldiers.
- Dr. Baker is an homage to OvenBaker, the contributor who added the Close Quarters Battle CBM and Bionic Combatives martial art to the game in #6334
- The Zombie Bio Operators found in game are the same soldiers that were sent to collect the initial combat data for Bionic Combatives. They are equipped with a malfunctioning, but still active, CQB CBM running Bionic Combatives. Their takedown special attack is based off of the Biojutsu Takedown technique.
- It's possible one of the supercomputers could still be active and awaiting combat data...
- There was a second, more powerful group of operators sent into the field with Bionic Combatives right before the Cataclysm occurred...

### CODE REVIEW

File: data/json/techniques.json [hunk 1]
```diff
@@ -275,6 +275,83 @@
...
+    "melee_allowed": true,
+    "crit_tec": true,
+    "mult_bonuses": [ [ "damage", "bash", 1.5 ], [ "damage", "cut", 1.5 ], [ "damage", "stab", 1.5 ] ],
+    "stun_dur": 1
+  },
+  {
+    "type": "technique",
+    "id": "tec_biojutsu_takedown",
+    "name": "Biojutsu Takedown",
+    "messages": [ "You kick and slam %s to the ground", "<npcname> kick and slam %s to the ground" ],
```
Comment (NickGrease) [inline data/json/techniques.json]:
```suggestion
    "messages": [ "You kick and slam %s to the ground", "<npcname> kicks and slams %s to the ground" ],
```

**Comment by Hymore246** [comment]
Jenkins rebuild

**Comment by Night-Pryanik** [comment]
I guess you could put the MA definition in `martialarts_fictional.json` instead of generic `martialarts.json`, as it's, you know, a fictional martial art.

**Comment by Hymore246** [comment]
Yeah, that makes sense. I moved it over so it can hangout with the Five Venoms.

File: data/json/martialarts_fictional.json [hunk 2]
```diff
@@ -1,4 +1,48 @@
 [
+  {
+    "type": "martial_art",
+    "id": "style_biojutsu",
+    "name": "Bionic Combatives",
+    "description": "A modern combat style for the post-modern human.  Nicknamed \"Biojutsu\", Bionic Combatives combines integrated weaponry, armor and augments into an consolidated fighting discipline.",
+    "initiate": [ "BEGINING BIONIC COMBATIVES PROGRAM. INITIATING COMBAT PROTOCOLS.", "%s starts moving with swift robotic percision." ],
```
Comment (Night-Pryanik) [inline data/json/martialarts_fictional.json]:
```suggestion
    "initiate": [ "BEGINNING BIONIC COMBATIVES PROGRAM. INITIATING COMBAT PROTOCOLS.", "%s starts moving with swift robotic precision." ],
```

**Comment by faefux** [comment]
Really appreciate you writing an interesting backstory for this, already implemented it in some dialogue! :)

**Comment by geophree** [comment]
Instead of Dr. Paul Baker, what about Dr. Owen Baker? (OvenBaker -> Owen Baker)

**Pull request merged**